sympy/logic/boolalg.py

```diff
@@ -628,34 +649,135 @@ def _apply_patternbased_simplification(self, rv, patterns, measure,
+                                # We only want to use ITE and Min/Max replacements if
+                                # they simplify to a relational
+                                costsaving = measure(func(*oldexpr.args)) - measure(np)
                                 if costsaving > 0:
-                                    results.append((costsaving, (i, j, np)))
+                                    results.append((costsaving, ([i, j], np)))
             if results:
                 # Sort results based on complexity
                 results = list(reversed(sorted(results,
                                                key=lambda pair: pair[0])))
                 # Replace the one providing most simplification
                 replacement = results[0][1]
-                i, j, newrel = replacement
+                idx, newrel = replacement
+                idx.sort()
                 # Remove the old relationals
-                del Rel[j]
-                del Rel[i]
+                for index in reversed(idx):
+                    del Rel[index]
                 if dominatingvalue is None or newrel != ~dominatingvalue:
                     # Insert the new one (no need to insert a value that will
                     # not affect the result)
-                    Rel.append(newrel)
+                    if newrel.func == func:
+                        for a in newrel.args:
+                            Rel.append(a)
+                    else:
+                        Rel.append(newrel)
                 # We did change something so try again
                 changed = True
+        return Rel
 
-        rv = rv.func(*([_canonical(i) for i in ordered(Rel)]
-                     + nonRel + nonRealRel))
-        return rv
+    def _apply_patternbased_threeterm_simplification(self, Rel, patterns, func,
+                                                     dominatingvalue,
+                                                     replacementvalue,
+                                                     measure):
+        from sympy.functions.elementary.miscellaneous import Min, Max
+        changed = True
+        while changed and len(Rel) >= 3:
+            changed = False
+            # Sort based on ordered
+            Rel = list(ordered(Rel))
+            # Create a list of possible replacements
+            results = []
+            # Try all combinations
+            for ((i, pi), (j, pj), (k, pk)) in permutations(enumerate(Rel), 3):
+                for pattern, simp in patterns:
+                    res = []
+                    # use SymPy matching
+                    oldexpr = Tuple(pi, pj, pk)
+                    tmpres = oldexpr.match(pattern)
+                    if tmpres:
+                        res.append((tmpres, oldexpr))
+                    # Try reversing first relational
+                    # This and the rest should not be required with a better
+                    # canonical
```

Comment:
I do not recall the details, but for example, `x < 1` and `1 > x` are both `x < 1` in canonical form, while `1 < x` is `x > 1`, so it is preferred that the symbol is on the left. From the perspective of simplification it is better if relationals were always written either as < or > as there would be fewer ways to test the matching for.

However, even so, for (un)equalities there will be a ordering of the variables, so it somehow these must match (or one need to at least reverse the (un)equalities). At the moment, the solution is to try all combinations of reversed relationals to see if they match the patterns.

Say `And(Lt(x, y), Eq(x, y))` will match a pattern, but not `And(Lt(x, y), Eq(y, x))`.

This can probably be better written so that inequalities are always in a single form (either less or greater) and then only the (un)equalities are reversed. However, I do not now really see how to write that in a compact, general, form.

Reply:
I'm working on a better version now...

Reply:
OK! I think I've got a version that actually tests the minimum number of required times (for some definition).

1) I use `combinations` to reduce the code size
2) Only Eq and Ne are reversed
3) Patterns that are now redundant are commented out.
4) Fixed some issues with previous patterns.